### ISSUE TITLE
Allow for more mappings in the I/O registry

### ISSUE THREAD

**Issue by gtca** [scverse/anndata#796]
Objects than contain for instance `OrderedDict` in `.uns` elements fail to be serialised since `anndata==0.8`:

```python
from collections import OrderedDict
import numpy as np
from jax import random
from anndata import AnnData

x = random.normal(key=random.PRNGKey(1), shape=(1000,500))
adata = AnnData(np.array(x))

adata.uns["ord_dict"] = OrderedDict({"key1": "value1"})
adata.write("anndata_with_ordered_dict.h5ad")
# => TypeError: No method has been defined for writing <class 'collections.OrderedDict'> elements to <class 'h5py._hl.group.Group'>
```

Currently users need to identify items in `.uns` that cause this error and convert them to `dict`.

While there's probably no reason to use `OrderedDict` in practice, it would be great to have support for it as well as for other dictionaries such as [`ConfigDict`](https://ml-collections.readthedocs.io/en/latest/config_dict.html).

**Comment by gtca** [comment]
The correct way would probably be checking if it's a mapping and converting it to a dict before writing it on disk.

```py
from collections import OrderedDict
od = OrderedDict(a = 1)

from collections.abc import Mapping
isinstance(od, Mapping)  # => True
```

Unfortunately, that might not work for all the third-party objects:

```py
from ml_collections import config_dict
cfg = config_dict.ConfigDict()

isinstance(cfg, Mapping)  # => False
```

**Comment by gtca** [comment]
@ivirshup do you think we could go ahead with `isinstance(item, Mapping)` for instance?

**Comment by ivirshup** [comment]
I don't think so. Mappings are a fairly common type in python, and many things which are mappings we wouldn't want to convert to a dict. I'd be okay with adding overloads for some mapping types for some cases from the standard library, like `OrderedDict`.

**Comment by gtca** [comment]
Sounds good enough to me as well!

**Comment by ivirshup** [comment]
Would you open a PR?

I think it should be pretty straight forward. Just adding more classes to the write_dict method plus tests.

**Comment by ivirshup** [comment]
On handling of ordered dict:

Right now I would lean towards just returning a dict. I would be up for maintaining the order of the keys (as much as possible) though. We can do this with `track_order` for hdf5. I don't think there's a native way to do this for zarr though.

What is the use case for OrderedDicts?

**Comment by ivirshup** [comment]
Also, this could be a bug fix if we're just not erroring for OrderedDicts. If we actually try to maintain order, I think we would need to update the schema to save the order in attributes.

I'd want some good use cases for that first.

It could potentially be done with an optional attribute `key-order`, which if defined determines the read order. But this would take some planning for release for compatibility.

**Comment by ktpolanski** [comment]
Ran into this by proxy today via muon. Dropping a line in here so I'm up to date with developments on the matter.

**Comment by flying-sheep** [comment]
I added a PR for this, it’s easy enough: #1008

**Comment by ivirshup** [comment]
It's very easy to add, but adding more types to serialize adds more work to other implementations. So I'd like to be conservative about this.

E.g. if we really need to maintain order (which I would like to see use cases for), maybe we should just do that with mappings, rather than adding a new type.

**Comment by flying-sheep** [comment]
OrderedDict doesn’t have many use cases these days, it doesn’t do much different than a regular one.

But it is used in older code that simply wants the insertion order that `dict` has today.

So encountering those OrderedDicts in the wild is relatively common.

That being said, do we serialize item order for dicts? If not, then my PR is probably not complete …

**Comment by ivirshup** [comment]
Encountering ordered dicts is kinda common, but idk that means we need to support round-tripping them inside and anndata. Like, maybe we could convert them to dicts on write. But then we'd want to try preserve order of elements for dicts.

**Comment by flying-sheep** [comment]
> But then we'd want to try preserve order of elements for dicts.

Maybe we should do that anyway. We’re preserving order for `obs` and `var`, why shouldn’t users be able to do the same.

And as said: `dict` now has stable order, so people come to rely on it. Would be nice if it would survive an IO roundtrip.
